File: 1218-longest-arithmetic-subsequence-of-given-difference/1218-longest-arithmetic-subsequence-of-given-difference.java
class Solution {
    public int longestSubsequence(int[] arr, int difference) {
        int ans = 0;
        var dp = new HashMap<Integer, Integer>();
        for(int i : arr) {
            if(dp.containsKey(i - difference))
                dp.put(i, 1 + dp.get(i - difference));
            else 
                dp.put(i, 1);
            ans = Math.max(ans, dp.get(i));
        }
        return ans;
    }
}

File: 1218-longest-arithmetic-subsequence-of-given-difference/1218-longest-arithmetic-subsequence-of-given-difference.cpp
class Solution {
public:
    int longestSubsequence(vector<int>& arr, int difference) {
        unordered_map<int, int> dp;
        int ans = 0;
        for(int &i : arr) {
            if(dp.count(i - difference))
                dp[i] = 1 + dp[i - difference];
            else dp[i] = 1;
            ans = max(ans, dp[i]);
        }
        return ans;
    }
};